### PR TITLE
Resolve circular references using id entry references

### PR DESCRIPTION
**Pull request by dv-msft**
The current 'map <--> contained-objects' relationship uses strong circular references maintained between object-maps (map-of-maps, hash-map-of-maps) and objects (programs and maps).

While this design works for well-behaved user mode applications, it breaks down if the user mode applications do not perform explicit clean-up via the provided clean-up APIs, leading to leaked resources by the way of programs and maps orphaned in memory. This typically happens with user programs that either exit prematurely, do not call the clean-up API calls or crash before they can do so.

This PR proposes a new design solution where we use references to the id table entry for object instead of the object itself. 

**Testing**

_User mode:_

Ensure that the ```unit_tests.exe``` test suite passes successfully.

_Kernel Mode:_

- Ensure that the ```api_test.exe``` test suite passes successfully, multiple times back-to-back.
- The ```epbfcore.sys``` driver can be unloaded and re-loaded successfully after a multiple pass test run.

_Testing Notes:_

Some edge cases were not covered by the existing tests, so new tests (user and kernel-mode) were added.
I'm seeing some inconsistencies in API behavior across different map types. My plan is to verify API functionality on linux for these scenarios and create new issues where required.

**Documentation**

No new documentation required.

Fixes: #1509 #1510

### CODE REVIEW

File: libs/execution_context/ebpf_maps.c [hunk 1]
```diff
@@ -618,52 +623,102 @@ _update_array_map_entry_with_handle(
     uintptr_t value_handle,
     ebpf_map_option_t option)
 {
-    if (!map || !key || (option == EBPF_NOEXIST))
+    ebpf_result_t result = EBPF_SUCCESS;
+
+    if (!map || !key || (option == EBPF_NOEXIST)) {
+        EBPF_LOG_MESSAGE(EBPF_TRACELOG_LEVEL_ERROR, EBPF_TRACELOG_KEYWORD_MAP, "Invalid argument(s) rejected");
         return EBPF_INVALID_ARGUMENT;
```
Comment (dv-msft) [inline libs/execution_context/ebpf_maps.c:630]: Working on adding additional tests to address code coverage issues (where possible)

File: libs/platform/ebpf_object.c [hunk 2]
```diff
@@ -11,12 +11,12 @@ static ebpf_lock_t _ebpf_object_tracking_list_lock = {0};
 /**
  * @brief Objects are added to the ID table when they are initialized and removed
  * from the table when they are freed. Objects in the table always have a
```
Comment (Alan-Jowett) [inline libs/platform/ebpf_object.c:13]: Not true anymore. Please fix this comment,

File: include/ebpf_result.h [hunk 3]
```diff
@@ -112,6 +112,9 @@ extern "C"
 
         /// Operation timed out.
         EBPF_TIMEOUT,
+
+        /// Key is valid, but the object has been deleted
```
Comment (dthaler) [inline include/ebpf_result.h:116]:
```suggestion
        /// Key is valid, but the object has been deleted.
```
Use style consistent with rest of defines above

File: libs/platform/ebpf_object.c [hunk 4]
```diff
@@ -37,6 +37,20 @@ typedef struct _ebpf_id_entry
     // Counter incremented each time a new object is stored here.
     uint16_t counter;
 
+    // Reference count for this entry itself. This will be 1 when the pointed-to object (below) is created and
+    // incremented for every other object that holds a reference to this entry.
+    //
+    // If the pointed-to object gets destroyed first:
+    // - the object pointer will be nulled out and the id entry ref count will be decremented. After this point,
+    //   other objects that continue to hold a reference to this id entry, will fail to acquire a reference to the
+    //   pointed-to object (via it's id) and will need to handle the failure gracefully.
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:46]:
```suggestion
    //   pointed-to object (via its id) and will need to handle the failure gracefully.
```

File: libs/platform/ebpf_object.c [hunk 5]
```diff
@@ -112,10 +127,12 @@ _Requires_lock_held_(&_ebpf_object_tracking_list_lock) static void _ebpf_object_
     ebpf_result_t return_value = _get_index_from_id(object->id, &index);
     ebpf_assert(return_value == EBPF_SUCCESS);
 
-    // In a release build, ebpf_assert is a no-op so we
-    // need to avoid an unreferenced variable warning.
+    // In a release build, ebpf_assert is a no-op so we need to avoid an unreferenced variable warning.
     UNREFERENCED_PARAMETER(return_value);
 
+    // Under lock, so un-protected access if ok.
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:133]:
```suggestion
    // Under lock, so un-protected access is ok.
```

File: libs/platform/ebpf_object.c [hunk 6]
```diff
@@ -368,3 +388,107 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+        }
+        goto Done;
+    }
+
+    // ref count is non-zero
+    if (entry->object == NULL) {
+
+        // The object at this entry has been deleted and all that remains are (now stale) references to this entry
+        // held by other objects.  This is a non-reversible situation and there's no point in giving out references
+        // for such entries, so deny with suitable error code.
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:425]:
```suggestion
        // for such entries, so deny with a suitable error code.
```

File: tests/api_test/api_test.cpp [hunk 7]
```diff
@@ -759,3 +784,95 @@ bpf_user_helpers_test(ebpf_execution_type_t execution_type)
...
+// This test tests resource reclamation and clean-up after a premature/abnormal user mode application exit.
+TEST_CASE("close_unload_test", "[native_tests][native_close_cleanup_tests]")
+{
+    struct bpf_object* object = nullptr;
+    hook_helper_t hook(EBPF_ATTACH_TYPE_BIND);
+    program_load_attach_helper_t _helper(
+        "bindmonitor_tailcall.sys", BPF_PROG_TYPE_BIND, "BindMonitor", EBPF_EXECUTION_NATIVE, nullptr, 0, hook);
+    object = _helper.get_object();
+
+    // Setup tail calls.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:797]:
```suggestion
    // Set up tail calls.
```

File: tests/api_test/api_test.cpp [hunk 8]
```diff
@@ -759,3 +784,95 @@ bpf_user_helpers_test(ebpf_execution_type_t execution_type)
...
+    // of reclaiming and free'ing such objects. All unit tests belonging to the '[native_close_cleanup_tests]'
+    // unit-test class simulate this behavior by _not_ calling the clean-up api calls.
+    //
+    // For native tests (meant for execution on the kernel mode ebpf-for-windows driver), this event will be handled
+    // by the ebpf-core kernel mode driver on test application termination.
+    //
+    // The success/failure of the [native_close_cleanup_tests] tests can only be (indirectly) checked by attempting to
+    // stop the ebpf-core driver after executing this class of tests.  If the clean-up by the ebpf-core driver is not
+    // successful, it cannot be stopped/unloaded.  This step is performed automatically by the CI/CD test pass runs and
+    // will need to be perfomed as an explicit manual step after a manually initiated test-run.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:843]: Is it still safe to run api_test manually multiple times in a row without stopping the ebpf core driver?

File: libs/execution_context/ebpf_maps.c [hunk 9]
```diff
@@ -444,33 +446,37 @@ _associate_inner_map(_Inout_ ebpf_core_object_map_t* object_map, ebpf_handle_t i
...
         ebpf_id_t id = *(ebpf_id_t*)&map->data[i * map->ebpf_map_definition.value_size];
         if (id) {
-            ebpf_assert_success(ebpf_object_dereference_by_id(id, value_type));
+            ebpf_assert_success(ebpf_object_release_id_reference(id, value_type));
         }
     }
 
-    if (object_map->inner_map_template != NULL) {
-        ebpf_object_release_reference(&object_map->inner_map_template->object);
-    }
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:472]: Why is it we don't need to release the template id here?

File: libs/execution_context/ebpf_maps.c [hunk 10]
```diff
@@ -618,52 +623,102 @@ _update_array_map_entry_with_handle(
...
-    ebpf_core_object_map_t* object_map = EBPF_FROM_FIELD(ebpf_core_object_map_t, core_map, map);
     ebpf_core_object_t* value_object = NULL;
-
     if (value_handle != (uintptr_t)ebpf_handle_invalid) {
         result = ebpf_object_reference_by_handle(value_handle, value_type, &value_object);
         if (result != EBPF_SUCCESS) {
+            EBPF_LOG_MESSAGE_UINT64_UINT64(
+                EBPF_TRACELOG_LEVEL_ERROR,
+                EBPF_TRACELOG_KEYWORD_MAP,
+                "Invalid Object handle rejected",
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:647]:
```suggestion
                "Invalid object handle rejected",
```

File: libs/execution_context/ebpf_maps.c [hunk 11]
```diff
@@ -1317,33 +1387,58 @@ _update_hash_map_entry_with_handle(
...
+    result = ebpf_object_acquire_id_reference(value_object->id, value_type);
+    ebpf_assert(result == EBPF_SUCCESS);
+    if (result != EBPF_SUCCESS) {
+
+        // This can _never_ happen as we already have a valid pointer to the object.  This is indicative of a
+        // fatal internal error.
+        EBPF_LOG_MESSAGE_UINT64_UINT64(
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_MAP,
+            "Invalid Object Id rejected",
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:1434]:
```suggestion
            "Invalid object ID rejected",
```

File: libs/platform/ebpf_object.c (hunk 2 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: Not sure which specific issue you were referring to, so expanded and added more details.  Will follow-up to make sure your concerns were addressed.

File: include/ebpf_result.h (hunk 3 above)
Comment (dv-msft) [reply include/ebpf_result.h]: fixed

File: libs/platform/ebpf_object.c (hunk 5 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: libs/platform/ebpf_object.c (hunk 6 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: tests/api_test/api_test.cpp (hunk 7 above)
Comment (dv-msft) [reply tests/api_test/api_test.cpp]: fixed

File: libs/execution_context/ebpf_maps.c (hunk 9 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c:472]: I'd noticed that we we were using the template map to only get its definition (to ensure each incoming map element's equivalence to the template map definition).  We now therefore store only the template map's definition (during the map-of-maps creation) and let go of the template map right then as I did not see any further dependence on the template map itself thereafter.

(I'd discussed this with @Alan-Jowett prior to going this route).

File: libs/execution_context/ebpf_maps.c (hunk 10 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c]: fixed

File: libs/execution_context/ebpf_maps.c (hunk 11 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c]: fixed

File: libs/platform/ebpf_object.c (hunk 4 above)
Comment (dv-msft) [reply libs/platform/ebpf_object.c]: fixed

File: tests/api_test/api_test.cpp (hunk 8 above)
Comment (dv-msft) [reply tests/api_test/api_test.cpp:843]: Yes. As a matter of fact, that was/is a part of my inner-loop testing.  I've updated the PR 'testing' section accordingly.

File: libs/execution_context/ebpf_maps.c (hunk 1 above)
Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c:634]: done

Comment (dv-msft) [reply libs/execution_context/ebpf_maps.c:634]: done

File: libs/platform/ebpf_object.c [hunk 12]
```diff
@@ -9,14 +9,25 @@ static const uint32_t _ebpf_object_marker = 'eobj';
 static ebpf_lock_t _ebpf_object_tracking_list_lock = {0};
 
 /**
- * @brief Objects are added to the ID table when they are initialized and removed
- * from the table when they are freed. Objects in the table always have a
- *  ref-count > 0.
+ * @brief Objects are allocated an entry in the the ID
+ * table when they are initialized.  Along with a pointer to
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:13]:
```suggestion
 * @brief Objects are allocated an entry in the ID
 *  table when they are initialized.  Along with a pointer to
```

File: libs/platform/ebpf_object.c [hunk 13]
```diff
@@ -368,3 +399,107 @@ ebpf_duplicate_string(_In_z_ const char* source)
...
+        goto Done;
+    }
+
+    // ref count is non-zero
+    if (entry->object == NULL) {
+
+        // The object at this entry has been deleted and all that remains are (the now stale) references to this entry
+        // held by other objects.  This is a non-reversible situation and there's no point in giving out references
+        // for such entries, so deny with a suitable error code.
+        result = EBPF_STALE_KEY;
```
Comment (dthaler) [inline libs/platform/ebpf_object.c:437]: Shouldn't this be called EBPF_STALE_ID?  Or is there any other type of key this is used for?

**Review by dthaler — APPROVED** [review]
Ok but see suggestions.

**Comment by dthaler** [comment]
Merging to fix the bug but please see non-blocking suggestions.
